### PR TITLE
feat: add tm_unslug() function to reverse tm_slug() from a dictionary

### PR DESCRIPTION
**Pull request by mcalhoun**
## What this PR does / why we need it:

This PR introduces the `tm_unslug()` function, which provides the inverse operation of `tm_slug()` by mapping slugs back to their original human-readable strings using a caller-provided dictionary.

### Key Features:
- **Polymorphic behavior**: Accepts both `string` and `list(string)` inputs, with output shape mirroring the input
- **Dictionary-based mapping**: Uses a provided list of unslugged canonical values to perform reliable reverse mapping
- **Smart normalization**: Handles trailing separators and collapsed internal separators for robust matching
- **Collision detection**: Raises a typed `UnslugErrorIndeterministic` error when a slug maps to multiple dictionary entries
- **Round-trip guarantee**: For all words in the dictionary: `tm_unslug(tm_slug(word), dictionary) == word`
- **Special character support**: Extensive support for Unicode, emoji, punctuation, and other special characters

### Implementation Details:
- Follows the established stdlib pattern with separate files for implementation, errors, and tests
- Comprehensive test coverage including edge cases, normalization, collisions, and special characters
- Proper error handling with typed errors for invalid inputs and indeterministic mappings

## Which issue(s) this PR fixes:

N/A - This is a new feature addition based on the product requirements document.

## Special notes for your reviewer:

1. The function normalizes input slugs to handle common variations (trailing separators, repeated separators) but intentionally does NOT strip leading separators to avoid false matches
2. The implementation precomputes a lookup map for O(1) lookups when processing lists
3. Extensive test cases cover special characters including Unicode, emoji, HTML entities, mathematical symbols, etc.
4. The function is registered in the stdlib registry alongside `tm_slug()` for consistency

## Does this PR introduce a user-facing change?

Yes - adds new stdlib function `tm_unslug()` that maps slugs back to original strings using a dictionary.

Example usage:
- Single string: `tm_unslug("team-alpha", ["Team Alpha", "Team Beta"]) => "Team Alpha"`
- List input: `tm_unslug(["foo-bar", "hello-world"], ["Foo Bar", "Hello World"]) => ["Foo Bar", "Hello World"]`
- No match fallback: `tm_unslug("unknown", ["Known"]) => "unknown"`
- Collision error: `tm_unslug("test", ["Test", "test"]) => error`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -50,6 +58,14 @@ Given a version number `MAJOR.MINOR.PATCH`, we increment the:
 
 - Add support for slashes (`/`) in tags.
 
+### Changed
```
Comment (snakster) [inline CHANGELOG.md]: This can be removed.

File: CHANGELOG.md [hunk 2]
```diff
@@ -20,10 +20,18 @@ Given a version number `MAJOR.MINOR.PATCH`, we increment the:
 - Backward compatibility in versions `0.0.z` is **not guaranteed** when `z` is increased.
 - Backward compatibility in versions `0.y.z` is **not guaranteed** when `y` is increased.
 
-## Unreleased
+## [Unreleased]
```
Comment (snakster) [inline CHANGELOG.md]: I think this one can be removed.

**Review by snakster — COMMENTED** [review]
LGTM, just two minor comments for the changelog.
